diff --git a/dxil_spirv.cpp b/dxil_spirv.cpp
index 06227db..5970c44 100644
--- a/dxil_spirv.cpp
+++ b/dxil_spirv.cpp
@@ -347,6 +347,7 @@ struct Remapper
 
 	bool ssbo_uav = false;
 	bool ssbo_srv = false;
+	bool class_bindings = false; // bbport: binding = register + 32 * class (SRV, UAV, CBV, sampler)
 	bool ssbo_rtas = false;
 	bool input_attachments = false;
 };
@@ -473,7 +474,7 @@ static dxil_spv_bool remap_sampler(void *userdata, const dxil_spv_d3d_binding *b
 	{
 		vk_binding->bindless.use_heap = DXIL_SPV_FALSE;
 		vk_binding->set = binding->register_space;
-		vk_binding->binding = binding->register_index;
+		vk_binding->binding = binding->register_index + (remapper->class_bindings ? 96u : 0u);
 	}
 	return DXIL_SPV_TRUE;
 }
@@ -512,7 +513,7 @@ static dxil_spv_bool remap_uav(void *userdata, const dxil_spv_uav_d3d_binding *b
 		{
 			vk_binding->buffer_binding.bindless.use_heap = DXIL_SPV_FALSE;
 			vk_binding->buffer_binding.set = binding->d3d_binding.register_space;
-			vk_binding->buffer_binding.binding = binding->d3d_binding.register_index;
+			vk_binding->buffer_binding.binding = binding->d3d_binding.register_index + (remapper->class_bindings ? 32u : 0u);
 		}
 
 		if (remapper->ssbo_uav)
@@ -599,7 +600,7 @@ static dxil_spv_bool remap_cbv(void *userdata, const dxil_spv_d3d_binding *bindi
 			{
 				vk_binding->vulkan.uniform_binding.bindless.use_heap = DXIL_SPV_FALSE;
 				vk_binding->vulkan.uniform_binding.set = binding->register_space;
-				vk_binding->vulkan.uniform_binding.binding = binding->register_index;
+				vk_binding->vulkan.uniform_binding.binding = binding->register_index + (remapper->class_bindings ? 64u : 0u);
 			}
 		}
 	}
@@ -813,6 +814,7 @@ int main(int argc, char **argv)
 	cbs.add("--bindless-cbv-as-ssbo", [&](CLIParser &) { args.bindless_cbv_as_ssbo = true; });
 	cbs.add("--ssbo-uav", [&](CLIParser &) { remapper.ssbo_uav = true; });
 	cbs.add("--ssbo-srv", [&](CLIParser &) { remapper.ssbo_srv = true; });
+	cbs.add("--class-bindings", [&](CLIParser &) { remapper.class_bindings = true; });
 	cbs.add("--ssbo-rtas", [&](CLIParser &) { remapper.ssbo_rtas = true; });
 	cbs.add("--input-attachments", [&](CLIParser &) { remapper.input_attachments = true; });
 	cbs.add("--ssbo-alignment", [&](CLIParser &parser) { args.ssbo_alignment = parser.next_uint(); });
